File: chess-app/src/engine/Game.js
import Board from './Board.js';

export default class Game {
    constructor() {
        this.board = new Board();
        this.currentTurn = 'white';
        this.gameOver = false;
        this.winner = null;
    }
    start(){
        this.board.setNewBoard();
        this.board.printBoard();
        return this.board.getBoard();
    }

    playMove(fromPos, toPos) {
        if (fromPos === toPos) {
            return this.board.getBoard();
        }
        const pieceAtFrom = this.board.getPieceAt(fromPos)
        if(pieceAtFrom === null){
            throw new Error('No piece at from position');
        }

        if(pieceAtFrom.color !== this.currentTurn){
            throw new Error('Wrong turn');
        }
        const pseudoLegalMoves = pieceAtFrom.getPseudoLegalMoves(this.board, fromPos)
        const isIncluded = pseudoLegalMoves.some(arr => JSON.stringify(arr) === JSON.stringify(toPos));
        if(!isIncluded){
            throw new Error('Not a legal move');
        }

        const legalMoves = this.getLegalMoves(pseudoLegalMoves, fromPos);
        const includedLegalMoves = legalMoves.some(arr => JSON.stringify(arr) === JSON.stringify(toPos));
        if (!includedLegalMoves) {
            throw new Error('Not a legal move');
        }

        // move the piece
        this._makeMove(this.board, fromPos, this.currentTurn, toPos);
        this.currentTurn = this.currentTurn === 'white' ? 'black' : 'white';
        
        const status = this.checkStatus();
        if (status) {
            console.log(status.status);
            if (status.status === 'checkmate') {
                console.log(`Checkmate! ${status.winner} wins!`);
            } else if (status.status === 'stalemate') {
                console.log('Stalemate! The game is a draw.');
            }
        }
        this.board.printBoard();
        return this.board.getBoard();
    }

    getLegalMoves(pseudoLegalMoves, fromPos) {
        const legalMoves = []
        pseudoLegalMoves.forEach((move)=>{
            const cloneBoard = this.board.deepCopy();
            this._makeMove(cloneBoard, fromPos, this.currentTurn, move);
            
            const isWhite = this.currentTurn === 'white';
            const enemyPieceList = isWhite ? cloneBoard.blackPieces : cloneBoard.whitePieces;
            const kingPos = isWhite ? cloneBoard.whiteKingPos : cloneBoard.blackKingPos;
            const check = cloneBoard.isCheck(enemyPieceList, kingPos)
            // const check = false
            if (!check) {
                legalMoves.push(move);
            }
        })

        return legalMoves;
    }

    getAllLegalMoves() {
        const legalMoves = []
        const allPseudoMoves = this.board.getAllMovesColor(this.currentTurn);
        for (const key of Object.keys(allPseudoMoves)) {
            if (legalMoves.length > 0) {
                return legalMoves;
            }

            let pseudoLegalMoves = allPseudoMoves[key]
            let fromPos = key.split(',').map(Number);

            legalMoves.push(...this.getLegalMoves(pseudoLegalMoves, fromPos))
        }
        return legalMoves;

    }

    isCheckmate() {
        const legalMoves = this.getAllLegalMoves();
        const enemyPieceList = this.currentTurn === 'white' ? this.board.blackPieces : this.board.whitePieces;
        const kingPos = this.currentTurn === 'white' ? this.board.whiteKingPos : this.board.blackKingPos;
        const check = this.board.isCheck(enemyPieceList, kingPos);

        return check && legalMoves.length === 0;
    }

    isStalemate() {
        const legalMoves = this.getAllLegalMoves();
        return legalMoves.length === 0;
    }

    checkStatus() {
        const legalMoves = this.getAllLegalMoves();
        const enemyPieceList = this.currentTurn === 'white' ? this.board.blackPieces : this.board.whitePieces;
        const kingPos = this.currentTurn === 'white' ? this.board.whiteKingPos : this.board.blackKingPos;
        const check = this.board.isCheck(enemyPieceList, kingPos);

        if( check && legalMoves.length === 0) {
            // this.gameOver = true;
            return { status: 'checkmate', winner: this.currentTurn === 'white' ? 'black' : 'white' };
        }
        if (legalMoves.length === 0) {
            // this.gameOver = true;
            return { status: 'stalemate', winner: null };
        }
        if (check) {
            return { status: 'check', winner: null };
        }
    }

    _makeMove(board, fromPos, currTurn, toPos) {
        const [fR, fC] = fromPos;
        const [tR, tC] = toPos;
        const isWhiteTurn = currTurn === 'white';
        const pieceAtFrom = board.getPieceAt(fromPos);
        const pieceAtTo = board.getPieceAt(toPos);
        const isKingMove = pieceAtFrom.type === 'king';
        const isPawnMove = pieceAtFrom.type === 'pawn';
        const isRookMove = pieceAtFrom.type === 'rook';
        const isRookCaptured = pieceAtTo !== null && pieceAtTo.type === 'rook';
        const isEnPassantMove = JSON.stringify(board.enPassantTarget) === JSON.stringify(toPos);
        const isCastling = isKingMove && (Math.abs(fC - tC) === 2)

        this._updatePiecePosition(board, fromPos, toPos, isWhiteTurn);
        this._updateCastlingRights(board, fromPos, toPos, currTurn, isKingMove, isRookMove, isRookCaptured);
        
        pieceAtFrom.hasMoved = true;
        board.board[tR][tC] = pieceAtFrom;
        board.board[fR][fC] = null;
        
        if (isPawnMove) {
            this._handlePawnMove(board, fR, tR, tC, isWhiteTurn, isEnPassantMove);
        }

        if (isKingMove) {
            this._updateKingPosition(board, toPos, isWhiteTurn);
        }

        if (isCastling) {
            const castleSide = tC > fC ? 'kingside' : 'queenside';
            this._updateCastlingRookPosition(board, currTurn, castleSide);
        }

        if (!isPawnMove || Math.abs(fR - tR) !== 2){
            board.enPassantTarget = null;
        }
        
    }

    _updatePiecePosition(board, fromPos, toPos, isWhiteTurn) {
        const pieceList = isWhiteTurn ? board.whitePieces : board.blackPieces;
        const opponentPieces = isWhiteTurn ? board.blackPieces : board.whitePieces;

        this._removeValue(pieceList, fromPos);
        pieceList.push(toPos);

        const pieceAtTo = board.getPieceAt(toPos);
        if (pieceAtTo) {
            this._removeValue(opponentPieces, toPos);
        }
    }

    _updateCastlingRights(board, fromPos, toPos, currTurn, isKingMove, isRookMove, isRookCaptured) {
        if (isKingMove) {
            board.castlingRights[currTurn].kingside = false;
            board.castlingRights[currTurn].queenside = false;
        } else if (isRookMove) {
            if (fromPos[0] === 7 && fromPos[1] === 7 && currTurn === 'white') {
                board.castlingRights[currTurn].kingside = false;
            } else if (fromPos[0] === 7 && fromPos[1] === 0 && currTurn === 'white') {
                board.castlingRights[currTurn].quuenside = false;
            } else if (fromPos[0] === 0 && fromPos[1] === 7 && currTurn === 'black') {
                board.castlingRights[currTurn].kingside = false;
            } else if (fromPos[0] === 0 && fromPos[1] === 0 && currTurn === 'black') {
                board.castlingRights[currTurn].queenside = false;
            }
        } else if (isRookCaptured) {
            if (toPos[0] === 7 && toPos[1] === 7 && currTurn === 'black') {
                board.castlingRights['white'].kingside = false;
            } else if (toPos[0] === 7 && toPos[1] === 0 && currTurn === 'black') {
                board.castlingRights['whtie'].quuenside = false;
            } else if (toPos[0] === 0 && toPos[1] === 7 && currTurn === 'white') {
                board.castlingRights['black'].kingside = false;
            } else if (toPos[0] === 0 && toPos[1] === 0 && currTurn === 'white') {
                board.castlingRights['black'].queenside = false;
            }
        }
    }

    _updateCastlingRookPosition(board, currTurn, castleSide) {
        const rookInitialPositions = {
            white: {kingside: [7, 7], queenside: [7, 0]},
            black: {kingside: [0, 7], queenside: [0, 0]}
        }
        const rookFinalPositions = {
            white: {kingside: [7, 5], queenside: [7, 3]},
            black: {kingside: [0, 5], queenside: [0, 3]}
        }
        const [fR, fC] = rookInitialPositions[currTurn][castleSide];
        const [tR, tC] = rookFinalPositions[currTurn][castleSide];
        const rookAtIP = board.getPieceAt([fR, fC])
        rookAtIP.hasMoved = true;
        board.board[tR][tC] = rookAtIP;
        board.board[fR][fC] = null; 
        this._updatePiecePosition(board, [fR, fC], [tR, tC], currTurn === 'white');
    }

    _handlePawnMove(board, fR, tR, tC, isWhiteTurn, isEnPassantMove) {
        if (isEnPassantMove) {
            const capturedPawnPos = isWhiteTurn ? [tR + 1, tC] : [tR - 1, tC];
            const opponentPieces = isWhiteTurn ? board.blackPieces : board.whitePieces;

            this._removeValue(opponentPieces, capturedPawnPos);
            board.board[capturedPawnPos[0]][capturedPawnPos[1]] = null;
        } else if (Math.abs(fR - tR) === 2) {
            board.enPassantTarget = isWhiteTurn ? [tR + 1, tC] : [tR - 1, tC];
        } else if (tR === 0 || tR === 7) {
            const promotedPiece =  'Queen'; // You can change this to any piece type
            board.setNewPiece(tR, tC, promotedPiece, isWhiteTurn ? 'white' : 'black');
        }
    }

    _updateKingPosition(board, toPos, isWhiteTurn) {
        if (isWhiteTurn) {
            board.whiteKingPos = toPos;
        } else {
            board.blackKingPos = toPos;
        }
    }

    _removeValue(array, value){
        const index = array.findIndex(arr => 
            arr.length === value.length && arr.every((val, i) => val === value[i])
        );
        if (index > -1) {
            array.splice(index, 1)
        }
        return array
    }

    getMoves(r,c){
        return this.board.getMoves(r,c)
    }
}

File: chess/src/main/java/com/app/chess/Game.java
package com.app.chess;

import java.util.List;

import org.apache.logging.log4j.util.Strings;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.HashMap;
import com.app.chess.Pieces.*;

public class Game {
    Board board;
    Player currentTurn;
    boolean gameOver;
    String winner;

    public Game() {
         board = new Board();
         currentTurn = Player.WHITE;
         gameOver = false;
         winner = null;
    }

    public String[][] start() {
        board.setNewBoard();
        board.printBoard();
        return  board.getBoard();
    }

    public String[][] playMove(int[] fromPos, int[] toPos) {
        // Check if source and destination are the same
        if (fromPos == toPos) {
            System.out.println("Invalid move: Source and destination are the same.");
            return board.getBoard();
        }

        // Get the piece at the source position
        Piece pieceAtFrom = board.getPieceAt(fromPos);
        
        // Check if there is a piece at the source position
        if (pieceAtFrom == null) {
            System.out.println("Invalid move: No piece at source position.");
            return board.getBoard();
        }
        
        // Check if the piece belongs to the current player
        if (pieceAtFrom.color != currentTurn) {
            System.out.println("Wrong turn: Piece does not belong to current player.");
            return board.getBoard();
        }

        // Check if the move is legal for the piece
        List<int[]> psuedoLegalMoves = pieceAtFrom.getPsuedoLegalMoves(board, fromPos);
        boolean isIncluded = false;
        for (int[] move : psuedoLegalMoves) {
            if (Arrays.equals(move, toPos)) {
                isIncluded = true;
                break;
            }
        }
        if (!isIncluded) {
            System.out.println("Invalid move: Move not allowed for this piece.");
            return board.getBoard();
        }

        // Check if the move puts own king in check
        List <int []> legalMoves = getLegalMoves(psuedoLegalMoves, fromPos);
        if (legalMoves.isEmpty()) {
            System.out.println("Invalid move: Move would put own king in check.");
            return board.getBoard();
        } else {
            boolean isLegal = false;
            for (int[] move : legalMoves) {
                if (Arrays.equals(move, toPos)) {
                    isLegal = true;
                    break;
                }
            }
            if (!isLegal) {
                System.out.println("Invalid move: Move would put own king in check.");
                return board.getBoard();
            }
        }

        // Make the move
        makeMove(board, fromPos, toPos, currentTurn);
        currentTurn = currentTurn == Player.WHITE ? Player.BLACK : Player.WHITE;

        HashMap<String, String> status = checkStatus();
        if (status != null) {
            System.out.println(status.get("status"));
            if (status.get("status").equals("checkmate")) {
                gameOver = true;
                winner = currentTurn == Player.WHITE ? "black" : "white";
            } else if (status.get("status").equals("stalemate")) {
                gameOver = true;
                winner = "draw";
            }
        }
        board.printBoard();
        return board.getBoard();
    }

    public List<int[]> getLegalMoves(List<int []> psuedoLegalMoves, int[] fromPos) {
        List<int[]> legalMoves = new ArrayList<>();
        for (int[] toPos : psuedoLegalMoves) {
            Board boardCopy = board.deepCopy();
            makeMove(boardCopy, fromPos, toPos, currentTurn);
            int[] kingPos = currentTurn == Player.WHITE ? boardCopy.whiteKingPos : boardCopy.blackKingPos;
            List<int []>enemyPieceList = currentTurn == Player.WHITE ? boardCopy.blackPieces : boardCopy.whitePieces;
            if (!boardCopy.isCheck(enemyPieceList, kingPos)) {
                legalMoves.add(toPos);
            }
        }
        return legalMoves;
    }

    public List<int []> getAllLegalMoves() {
        List<int []> allLegalMoves = new ArrayList<>();
        HashMap<int[], List<int[]>> allPseudoMoves = board.getAllMovesColor(currentTurn);
        for (int[] pos : allPseudoMoves.keySet()) {
            if (allLegalMoves.size() > 0) {
                return allLegalMoves;
            }

            List<int []> pseudoMoves = allPseudoMoves.get(pos);
            allLegalMoves.addAll((getLegalMoves(pseudoMoves, pos)));
        }
        return allLegalMoves;
    }

    public HashMap<String, String> checkStatus() {
        List <int[]> legalMoves = getAllLegalMoves();
        List <int[]> enemyPieceList = currentTurn == Player.WHITE ? board.blackPieces : board.whitePieces;
        int[] kingPos = currentTurn == Player.WHITE ? board.whiteKingPos : board.blackKingPos;
        boolean inCheck = board.isCheck(enemyPieceList, kingPos);

        if (inCheck && legalMoves.isEmpty()) {
            return new HashMap<>() {{
                put("status", "checkmate");
                put("winner", currentTurn == Player.WHITE ? "black" : "white");
            }};
        } else if (!inCheck && legalMoves.isEmpty()) {
            return new HashMap<>() {{
                put("status", "stalemate");
                put("winner", "draw");
            }};
        } else if (inCheck) {
            return new HashMap<>() {{
                put("status", "check");
                put("winner", Strings.EMPTY);
            }};
        }
        return null;
    }

    private void makeMove(Board board, int[] fromPos, int[] toPos, Player currentTurn) {
        int fR, fC, tR, tC;
        fR = fromPos[0]; fC = fromPos[1];
        tR = toPos[0]; tC = toPos[1];

        boolean isWhiteTurn = currentTurn == Player.WHITE;
        Piece pieceAtFrom = board.getPieceAt(fromPos);
        Piece pieceAtTo = board.getPieceAt(toPos);
        boolean isKingMove = pieceAtFrom.type == PieceType.KING;
        boolean isPawnMove = pieceAtFrom.type == PieceType.PAWN;
        boolean isRookMove = pieceAtFrom.type == PieceType.ROOK;
        boolean isRookCaptured = pieceAtTo != null && pieceAtTo.type == PieceType.ROOK;
        boolean inEnPassantMove = Arrays.equals(board.enPassantTarget, toPos);
        boolean isCastling = isKingMove && Math.abs(tC - fC) == 2;

        updatePiecePosition(board, fromPos, toPos, isWhiteTurn);
        updateCastlingRights(board, fromPos, toPos, currentTurn, isKingMove, isRookMove, isRookCaptured);

        pieceAtFrom.hasMoved = true;
        board.board[tR][tC] = pieceAtFrom;
        board.board[fR][fC] = null;
        
        // Handle special moves
        if (isPawnMove) {
            handlePawnMove(board, fR, fC, tR, tC, isWhiteTurn, inEnPassantMove);
        }
        if (isKingMove) {
            updateKingPosition(board, toPos, isWhiteTurn);
        }
        if (isCastling) {
            String castleSide = tC == 6 ? "kingside" : "queenside";
            updateCastlingRookPosition(board, currentTurn, castleSide);
        }
        if (!isPawnMove || Math.abs(tR - fR) != 2) {
            board.enPassantTarget = null;
        } 
    }

    private void updatePiecePosition(Board board, int[] fromPos, int[] toPos, boolean isWhiteTurn) {
        List<int[]> pieceList, opponentPieceList;
        if (isWhiteTurn) {
            pieceList = board.whitePieces;
            opponentPieceList = board.blackPieces;
        } else {
            pieceList = board.blackPieces;
            opponentPieceList = board.whitePieces;
        }

        removeValue(pieceList, fromPos);
        pieceList.add(toPos);

        Piece pieceAtTo = board.getPieceAt(toPos);
        if (pieceAtTo != null) {
            removeValue(opponentPieceList, toPos);
        }
    }

    private void updateCastlingRights(Board board, int[] fromPos, int[] toPos, Player currTurn, boolean isKingMove, 
                                     boolean isRookMove, boolean isRookCaptured) {
        
        if (isKingMove) {
            board.castlingRights.get(currTurn).put("kingside", false);
            board.castlingRights.get(currTurn).put("queenside", false);
        } else if (isRookMove) {
            if (fromPos[0] == 7 && fromPos[1] == 7 && currTurn == Player.WHITE) {
                board.castlingRights.get(currTurn).put("kingside", false);
            } else if (fromPos[0] == 7 && fromPos[1] == 0 && currTurn == Player.WHITE) {
                board.castlingRights.get(currTurn).put("queenside", false);
            } else if (fromPos[0] == 0 && fromPos[1] == 7 && currTurn == Player.BLACK) {
                board.castlingRights.get(currTurn).put("kingside", false);
            } else if (fromPos[0] == 0 && fromPos[1] == 0 && currTurn == Player.BLACK) {
                board.castlingRights.get(currTurn).put("queenside", false);
            }
        } else if (isRookCaptured) {
            if (toPos[0] == 7 && toPos[1] == 7 && currTurn == Player.BLACK) {
                board.castlingRights.get(Player.WHITE).put("kingside", false);
            } else if (toPos[0] == 7 && toPos[1] == 0 && currTurn == Player.BLACK) {
                board.castlingRights.get(Player.WHITE).put("queenside", false);
            } else if (toPos[0] == 0 && toPos[1] == 7 && currTurn == Player.WHITE) {
                board.castlingRights.get(Player.BLACK).put("kingside", false);
            } else if (toPos[0] == 0 && toPos[1] == 0 && currTurn == Player.WHITE) {
                board.castlingRights.get(Player.BLACK).put("queenside", false);
            }
        }                                
    }

    private void handlePawnMove(Board board, int fR, int fC, int tR, int tC, boolean isWhiteTurn, boolean inEnPassantMove) {
        if (inEnPassantMove) {
            int direction = isWhiteTurn ? 1 : -1;
            int [] capturedPawnPos = {tR + direction, tC};
            List<int []> opponentPieceList = isWhiteTurn ? board.blackPieces : board.whitePieces;

            removeValue(opponentPieceList, capturedPawnPos);
            board.board[tR + direction][tC] = null;
        } else if (Math.abs(tR - fR) == 2) {
            board.enPassantTarget = isWhiteTurn ? new int[]{tR + 1, tC} : new int[]{tR - 1, tC};
        } else if (tR == 0 || tR == 7) {
            board.setNewPiece(tR, tC, "Queen", isWhiteTurn ? Player.WHITE : Player.BLACK);
        }
    }

    private void updateKingPosition(Board board, int[] toPos, boolean isWhiteTurn) {
        if (isWhiteTurn) {
            board.whiteKingPos = toPos;
        } else {
            board.blackKingPos = toPos;
        }
    }

    private void updateCastlingRookPosition(Board board, Player currTurn, String castleSide) {
        HashMap<Player, HashMap<String, int[]>> rookInitialPositions = new HashMap<>() {{
            put(Player.WHITE, new HashMap<>() {{
                put("kingside", new int[]{7, 7});
                put("queenside", new int[]{7, 0});
            }});
            put(Player.BLACK, new HashMap<>() {{
                put("kingside", new int[]{0, 7});
                put("queenside", new int[]{0, 0});
            }});
        }};

        HashMap<Player, HashMap<String, int[]>> rookFinalPositions = new HashMap<>() {{
            put(Player.WHITE, new HashMap<>() {{
                put("kingside", new int[]{7, 5});
                put("queenside", new int[]{7, 3});
            }});
            put(Player.BLACK, new HashMap<>() {{
                put("kingside", new int[]{0, 5});
                put("queenside", new int[]{0, 3});
            }});
        }};

        int[] rookFromPos = rookInitialPositions.get(currTurn).get(castleSide);
        int[] rookToPos = rookFinalPositions.get(currTurn).get(castleSide);
        Piece rook = board.getPieceAt(rookFromPos);
        rook.hasMoved = true;
        board.board[rookToPos[0]][rookToPos[1]] = rook;
        board.board[rookFromPos[0]][rookFromPos[1]] = null;
        updatePiecePosition(board, rookFromPos, rookToPos, currTurn == Player.WHITE);
    }

    private void removeValue(List<int[]> list, int[] value) {
        list.removeIf(pos -> Arrays.equals(pos, value));
    }

    public List<int[]> getMoves(int r, int c){
        return board.getMoves(r, c);
    }

}
